home_audit.ts

```diff
@@ -0,0 +1,691 @@
+const HesJsonSchema = require('./schema/hescore_json.schema.js');
+const Ajv = require("ajv");
+const addFormats = require('ajv-formats');
+const ajv = new Ajv({allErrors: true, strictTypes: false, strictSchema: false})
+addFormats(ajv);
+// Add the schema to the validator.
+ajv.addSchema(HesJsonSchema);
+// Add the custom keyword "error_msg" to the validator
+ajv.addKeyword('error_msg');
+
+const nullOrUndefined = [null, undefined];
+
+const mandatoryMessage = "Missing value for mandatory field";
+
+module.exports = getNestedValidationMessages;
+
+/**
+ * Perform the HES validation for the nested JSON format. Uses the JSON Schema for initial required field and field
+ * limit validations (e.g. enums, within bounds, etc) and then performs secondary cross validation across the building
+ * as a whole (e.g. the roof area is large enough to cover the floor area)
+ * @param {object} homeValues - JSON object which follows the HES Nested JSON Schema
+ * @return {{blocker: [{string, [string]}], error: [{string, [string]}], warning: [{string, [string]}]}} - Error messages
+ * for the homeValues object, grouped by severity (blocker, error, warning). Messages in each severity group are grouped by
+ * path in the JSON Schema to the error, and can contain multiple errors for a single item in the JSON Schema.
+ */
+let _errorMessages = {};
+function getNestedValidationMessages (homeValues) {
+    _errorMessages = {}
+
+    if(!ajv.validate(HesJsonSchema, homeValues)){
+        ajv.errors.forEach((error) => {
+            const {instancePath, params} = error;
+            const errorPath = params.missingProperty ? `${instancePath}/${params.missingProperty}` : instancePath;
+            addErrorMessage(errorPath, convertAJVError(error))
+        });
+    }
+    getCrossValidationMessages(homeValues);
+    return _errorMessages
+}
+
+
+/**
+ * Convert the AJV error into an intelligible error message that the HES system knows how to display
+ * @param {object} errorObj
+ * @return {string|undefined}
+ */
+function convertAJVError(errorObj) {
+    const {keyword, message, schemaPath} = errorObj;
+    const keyArr = schemaPath.split('/');
+    keyArr.shift(); // remove '#'
```

Comment:
Do we want a check on this to be sure?  `if (keyArr.startsWith("#")) => keyArr.shift()`